core/ipc/sink/off-heap/src/main/java/org/opennms/core/ipc/sink/offheap/QueueFileOffHeapDispatchQueue.java

```diff
@@ -0,0 +1,560 @@
+
+    private final Function<T, byte[]> serializer;
+    private final Function<byte[], T> deserializer;
+    private final String moduleName;
+    private final BlockingQueue<Map.Entry<String, T>> inMemoryQueue;
+
+    // Note the queue is not thread safe so access should be synchronized
+    private final QueueFile offHeapQueue;
+    private final long maxFileSizeInBytes;
+    private final int batchSize;
+    private final Batch batch;
+
+    private final ForkJoinPool serdesPool = new ForkJoinPool(
+            Math.max(Runtime.getRuntime().availableProcessors() - 1, 1));
+
+    // Used to guard access to the offHeapQueue and batch data structures
+    private final Lock offHeapLock = new ReentrantLock(true);
+    // Used to ensure only one thread can be enqueing at a time
+    private final Lock enqueueLock = new ReentrantLock(true);
+    private final FileCapacityLatch fileCapacityLatch = new FileCapacityLatch();
+
+    private final Method usedBytesMethod;
+
+    public QueueFileOffHeapDispatchQueue(Function<T, byte[]> serializer, Function<byte[], T> deserializer,
+                                         String moduleName, Path filePath, int inMemoryQueueSize, int batchSize,
+                                         long maxFileSizeInBytes) throws IOException {
+        Objects.requireNonNull(serializer);
+        Objects.requireNonNull(deserializer);
+        Objects.requireNonNull(moduleName);
+
+        if (inMemoryQueueSize < 1) {
+            throw new IllegalArgumentException("In memory queue size must be greater than 0");
+        }
+
+        if (inMemoryQueueSize % batchSize != 0) {
+            throw new IllegalArgumentException("In memory queue size must be a multiple of batch size");
+        }
+
+        if (maxFileSizeInBytes < 0) {
+            throw new IllegalArgumentException("Max file size must be either 0 or a positive integer");
+        }
+
+        this.serializer = serializer;
+        this.deserializer = deserializer;
+        this.moduleName = moduleName;
+        this.batchSize = batchSize;
+        this.maxFileSizeInBytes = maxFileSizeInBytes;
+        batch = new Batch(batchSize);
+
+        inMemoryQueue = new ArrayBlockingQueue<>(inMemoryQueueSize, true);
+
+        // Setting the max file size to 0 or less will disable the off-heap portion of this queue
+        if (maxFileSizeInBytes > 0) {
+            Objects.requireNonNull(filePath);
+            File file = Paths.get(filePath.toString(), moduleName + FILE_EXTENSION).toFile();
+
+            QueueFile qf;
+            try {
+                qf = new QueueFile.Builder(file).build();
+            } catch (Exception e) {
+                LOG.warn("Exception while loading queue file", e);
+
+                if (file.delete()) {
+                    qf = new QueueFile.Builder(file).build();
+                } else {
+                    throw new IOException("Could delete corrupted queue file " + file.getAbsolutePath());
```

Comment:
I think the exception should be  `Could not delete`

Reply:
Yep it should be.